purchases/src/main/kotlin/com/revenuecat/purchases/CustomerInfoHelper.kt

```diff
@@ -121,14 +125,30 @@ internal class CustomerInfoHelper(
             appInBackground,
             { info ->
                 log(LogIntent.RC_SUCCESS, CustomerInfoStrings.CUSTOMERINFO_UPDATED_FROM_NETWORK)
+                offlineEntitlementsManager.resetOfflineCustomerInfoCache()
                 cacheCustomerInfo(info)
                 sendUpdatedCustomerInfoToDelegateIfChanged(info)
                 dispatch { callback?.onReceived(info) }
             },
-            { error, _ ->
+            { error, isServerError ->
                 Log.e("Purchases", "Error fetching customer data: $error")
                 deviceCache.clearCustomerInfoCacheTimestamp(appUserID)
-                dispatch { callback?.onError(error) }
+                if (shouldCalculateOfflineCustomerInfo(isServerError, appUserID)) {
+                    offlineEntitlementsManager.calculateAndCacheOfflineCustomerInfo(
+                        appUserID,
+                        onSuccess = { offlineComputedCustomerInfo ->
+                            // TODO Improve logs
+                            warnLog("Using offline computed customer info.")
+                            dispatch { callback?.onReceived(offlineComputedCustomerInfo) }
+                        },
+                        onError = { errorCalculatingOfflineCustomerInfo ->
+                            errorLog(errorCalculatingOfflineCustomerInfo)
+                            dispatch { callback?.onError(error) }
+                        }
+                    )
+                } else {
```

Comment:
I thought about also disabling offline entitlements if we have a non-server error to try to go back to the normal flow as soon as possible. However, this could potentially be a network error or something else, which would mean that if the user enters offline entitlements mode and then loses connection, they would lose any granted entitlements so in the end I decided not to.

Reply:
so just to check, we're keeping it on as long as we get 4xx and 5xx, but 2xx and 3xx deactivate, right?

Reply:
That's correct! We will keep it enabled until we get a successful response from the backend. Any 4xx, 5xx or network errors won't disable it.